teamcity: failed test: TestCleanupTxnIntentsOnGCAsync

The following tests appear to have failed on master (test): TestCleanupTxnIntentsOnGCAsync, TestCleanupTxnIntentsOnGCAsync/#01, TestCleanupTxnIntentsOnGCAsync/#00, TestCleanupTxnIntentsOnGCAsync/#03, TestCleanupTxnIntentsOnGCAsync/#02

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+TestCleanupTxnIntentsOnGCAsync).

[#1160310](https://teamcity.cockroachdb.com/viewLog.html?buildId=1160310):


```
TestCleanupTxnIntentsOnGCAsync/#02
--- FAIL: test/TestCleanupTxnIntentsOnGCAsync/#02 (0.000s)
Test ended in panic.

------- Stdout: -------
W190303 18:44:35.806826 49 storage/intentresolver/intent_resolver.go:729  failed to cleanup transaction intents: failed to resolve intents: boom



TestCleanupTxnIntentsOnGCAsync
--- FAIL: testrace/TestCleanupTxnIntentsOnGCAsync (0.040s)
Test ended in panic.




TestCleanupTxnIntentsOnGCAsync/#00
--- FAIL: test/TestCleanupTxnIntentsOnGCAsync/#00 (0.000s)
Test ended in panic.




TestCleanupTxnIntentsOnGCAsync
--- FAIL: test/TestCleanupTxnIntentsOnGCAsync (0.000s)
Test ended in panic.




TestCleanupTxnIntentsOnGCAsync/#03
...
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:190 +0x3e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc000504120, 0x11034c0, 0xc0000ba010, 0xc000343050, 0x22, 0x0, 0x0, 0xc00052a0f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:325 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:320 +0x134

goroutine 47 [select]:
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run(0xc00052ea90, 0x11034c0, 0xc0000ba010)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:373 +0x2fa
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run-fm(0x11034c0, 0xc0000ba010)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:190 +0x3e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc000504120, 0x11034c0, 0xc0000ba010, 0xc000343110, 0x22, 0x0, 0x0, 0xc00052a170)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:325 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:320 +0x134

goroutine 48 [select]:
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run(0xc00052eb60, 0x11034c0, 0xc0000ba010)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:373 +0x2fa
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run-fm(0x11034c0, 0xc0000ba010)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:190 +0x3e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc000504120, 0x11034c0, 0xc0000ba010, 0xc0003431a0, 0x22, 0x0, 0x0, 0xc00052a180)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:325 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:320 +0x134

goroutine 59 [select]:
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run(0xc0000b48f0, 0x11034c0, 0xc0000ba010)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:373 +0x2fa
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run-fm(0x11034c0, 0xc0000ba010)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:190 +0x3e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc000504120, 0x11034c0, 0xc0000ba010, 0xc0005a4030, 0x22, 0x0, 0x0, 0xc00049c0c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:325 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:320 +0x134

goroutine 60 [select]:
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run(0xc0000b4d00, 0x11034c0, 0xc0000ba010)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:373 +0x2fa
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).run-fm(0x11034c0, 0xc0000ba010)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:190 +0x3e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc000504120, 0x11034c0, 0xc0000ba010, 0xc0005a4060, 0x22, 0x0, 0x0, 0xc00049c0d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:325 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:320 +0x134



TestCleanupTxnIntentsOnGCAsync/#01
--- FAIL: test/TestCleanupTxnIntentsOnGCAsync/#01 (0.000s)
Test ended in panic.





```

Please assign, take a look and update the issue accordingly.

I've hit a flake on this test several times:

https://teamcity.cockroachdb.com/viewLog.html?buildId=1174203&buildTypeId=Cockroach_UnitTests_Test&tab=buildLog&_focus=349993
```
panic: interface conversion: roachpb.Response is *roachpb.GCResponse, not *roachpb.ResolveIntentRangeResponse [recovered]
	panic: interface conversion: roachpb.Response is *roachpb.GCResponse, not *roachpb.ResolveIntentRangeResponse

goroutine 63 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0003aa1b0, 0x1105fc0, 0xc000034108)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x11f
panic(0xe79480, 0xc00019f1d0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/storage/intentresolver.(*IntentResolver).ResolveIntents(0xc00051c680, 0x1105f80, 0xc000040e00, 0xc000360540, 0x3, 0x3, 0x410001, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intentresolver/intent_resolver.go:945 +0xfa1
github.com/cockroachdb/cockroach/pkg/storage/intentresolver.(*IntentResolver).cleanupFinishedTxnIntents(0xc00051c680, 0x1105fc0, 0xc000034108, 0x1, 0xc000313940, 0xc000360540, 0x3, 0x3, 0x158b4bf7b338dbe5, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intentresolver/intent_resolver.go:808 +0xed
github.com/cockroachdb/cockroach/pkg/storage/intentresolver.(*IntentResolver).CleanupTxnIntentsOnGCAsync.func1(0x1105fc0, 0xc000034108)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intentresolver/intent_resolver.go:726 +0x205
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunLimitedAsyncTask.func1(0xc0003aa1b0, 0x1105fc0, 0xc000034108, 0xf95af4, 0x16, 0xc00032f0e0, 0x0, 0x0, 0xc000286280)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:385 +0x110
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunLimitedAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:379 +0x23a
```

--- a/pkg/storage/intentresolver/intent_resolver_test.go
+++ b/pkg/storage/intentresolver/intent_resolver_test.go
@@ -94,7 +94,7 @@ func TestCleanupTxnIntentsOnGCAsync(t *testing.T) {
 	type testCase struct {
 		txn           *roachpb.Transaction
 		intents       []roachpb.Intent
-		sendFuncs     []sendFunc
+		sendFuncs     *sendFuncs
 		expectPushed  bool
 		expectSucceed bool
 	}
@@ -115,14 +115,17 @@ func TestCleanupTxnIntentsOnGCAsync(t *testing.T) {
 	txn2.Status = roachpb.COMMITTED
 	cases := []*testCase{
 		// This one has an unexpired pending transaction so it's skipped
-		{txn: txn0},
+		{
+			txn:       txn0,
+			sendFuncs: newSendFuncs(t),
+		},
 		// Txn1 is pending and expired so the code should attempt to push the txn.
 		// The provided sender will fail on the first request. The callback should
 		// indicate that the transaction was pushed but that the resolution was not
 		// successful.
 		{
 			txn:          txn1,
-			sendFuncs:    []sendFunc{failSendFunc},
+			sendFuncs:    newSendFuncs(t, failSendFunc),
 			expectPushed: true,
 		},
 		// Txn1 is pending and expired so the code should attempt to push the txn
@@ -143,11 +146,11 @@ func TestCleanupTxnIntentsOnGCAsync(t *testing.T) {
 					Txn:  txn1.TxnMeta,
 				},
 			},
-			sendFuncs: []sendFunc{
+			sendFuncs: newSendFuncs(t,
 				singlePushTxnSendFunc(t),
 				resolveIntentsSendFunc(t),
 				failSendFunc,
-			},
+			),
 			expectPushed: true,
 		},
 		// Txn1 is pending and expired so the code should attempt to push the txn
@@ -164,12 +167,15 @@ func TestCleanupTxnIntentsOnGCAsync(t *testing.T) {
 				{Span: roachpb.Span{Key: roachpb.Key("aa")}, Txn: txn1.TxnMeta},
 				{Span: roachpb.Span{Key: key, EndKey: roachpb.Key("b")}, Txn: txn1.TxnMeta},
 			},
-			sendFuncs: []sendFunc{
-				singlePushTxnSendFunc(t),
-				resolveIntentsSendFunc(t),
-				resolveIntentsSendFunc(t),
-				gcSendFunc(t),
-			},
+			sendFuncs: func() *sendFuncs {
+				s := newSendFuncs(t)
+				s.pushFrontLocked(
+					singlePushTxnSendFunc(t),
+					resolveIntentsSendFuncs(s, 3, 2),
+					gcSendFunc(t),
+				)
+				return s
+			}(),
 			expectPushed:  true,
 			expectSucceed: true,
 		},
@@ -179,15 +185,14 @@ func TestCleanupTxnIntentsOnGCAsync(t *testing.T) {
 		{
 			txn:           txn2,
 			intents:       []roachpb.Intent{},
-			sendFuncs:     []sendFunc{gcSendFunc(t)},
+			sendFuncs:     newSendFuncs(t, gcSendFunc(t)),
 			expectSucceed: true,
 		},
 	}
 
 	for _, c := range cases {
 		t.Run("", func(t *testing.T) {
-			sf := &sendFuncs{sendFuncs: c.sendFuncs}
-			ir := newIntentResolverWithSendFuncs(cfg, sf)
+			ir := newIntentResolverWithSendFuncs(cfg, c.sendFuncs)
 			var didPush, didSucceed bool
 			done := make(chan struct{})
 			onComplete := func(pushed, succeeded bool) {
@@ -199,7 +204,7 @@ func TestCleanupTxnIntentsOnGCAsync(t *testing.T) {
 				t.Fatalf("unexpected error sending async transaction")
 			}
 			<-done
-			if sf.len() != 0 {
+			if c.sendFuncs.len() != 0 {
 				t.Errorf("Not all send funcs called")
 			}
 			if didSucceed != c.expectSucceed {

